wsd/Storage.hpp

```diff
@@ -561,4 +513,47 @@ class LocalStorage : public StorageBase
     static std::atomic<unsigned> LastLocalStorageId;
 };
 
+/// Represents whether the underlying file is locked
+/// and with what token.
+struct LockContext
+{
+    /// Do we have support for locking for a storage.
+    bool _supportsLocks;
+    /// Do we own the (leased) lock currently
+    StorageBase::LockState _lockState;
+    /// Name if we need it to use consistently for locking
+    std::string _lockToken;
+    /// Time of last successful lock (re-)acquisition
+    std::chrono::steady_clock::time_point _lastLockTime;
+    /// Reason for unsuccessful locking request
+    std::string _lockFailureReason;
+
+    LockContext()
+        : _supportsLocks(false)
+        , _lockState(StorageBase::LockState::UNLOCK)
+        , _refreshSeconds(COOLWSD::getConfigValue<int>("storage.wopi.locking.refresh", 900))
+    {
+    }
+
+    /// one-time setup for supporting locks & create token
+    void initSupportsLocks();
+
+    /// Returns true if locks are supported.
+    bool supportsLocks() const { return _supportsLocks; }
+
+    /// Returns true if locked.
+    bool isLocked() const { return _lockState == StorageBase::LockState::LOCK; }
+
+    /// wait another refresh cycle
+    void bumpTimer() { _lastLockTime = std::chrono::steady_clock::now(); }
+
+    /// do we need to refresh our lock ?
+    bool needsRefresh(const std::chrono::steady_clock::time_point& now) const;
+
+    void dumpState(std::ostream& os) const;
+
+private:
+    const std::chrono::seconds _refreshSeconds;
+};
+
```

Comment:
I miss synchronization here as locking is usually implying multiple threads concurrently acting. If so, either we would need atomics (best for non-blocking locks) or at least thread locks (perhaps with blocking and waiting condition_variable etc)? @Ashod

Reply:
As stated above, this has nothing to do with threads. It's part of an external storage protocol to lock the document until the last editor disconnects.